jets/e/secp.c

```diff
@@ -0,0 +1,356 @@
+    fprintf(stderr, "\rsecp jet: crypto package error\n");
+    return u3m_bail(c3__exit);
+  }
+
+  c3_y s[32];
+  c3_y r[32];
+  byte_reverse(output64,      r, 32);
+  byte_reverse(output64 + 32, s, 32);
+
+  // package s,r,v signature for return
+  v += 27;
+  u3_noun s_noun  = u3i_words(8, (const c3_w*) s);
+  u3_noun r_noun  = u3i_words(8, (const c3_w*) r);
+  return (u3nt(v, r_noun, s_noun));
+
+}
+
+
+// recover pubkey from signature (which is how we verify signatures) 
+//
+
+u3_noun
+u3we_reco(u3_noun cor)
+{
+  u3_noun has, sig_v, sig_r, sig_s;
+  
+  if ( (c3n == u3r_mean(cor,
+                        u3x_sam_2,  &has,
+                        u3x_sam_6,  &sig_v,
+                        u3x_sam_14,  &sig_r,
+                        u3x_sam_15,  &sig_s,
+                        0)) ||
+       (c3n == u3ud(has)) ||
+       (c3n == u3ud(sig_v)) ||
+       (c3n == u3ud(sig_r)) ||
+       (c3n == u3ud(sig_s)) )
+    {
+      fprintf(stderr, "\rsecp jet: crypto package error\n");
+      return u3m_bail(c3__exit);
+    } else {
+    return u3qe_reco(has, sig_v, sig_r, sig_s);
+  }
+
+
+}
+
+u3_noun
+u3qe_reco(u3_atom hhas,
+          u3_atom sig_v,
+          u3_atom sig_r,
+          u3_atom sig_s)
+{
+  
+  // build library context object once (and only once)
+  static secp256k1_context * context = NULL;
+  if (NULL == context){
+    context = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY);
+  } 
+
+  // parse arguments, convert endianness
+  c3_y has[32];
+  c3_y   r[32];
+  c3_y   s[32];
+  c3_y   v[1];
+  u3r_bytes_reverse(0, 32, has, hhas);
+  u3r_bytes_reverse(0, 32, r, sig_r);
+  u3r_bytes_reverse(0, 32, s, sig_s);
+  u3r_bytes_reverse(0, 1,  v, sig_v);
+
+  // build the signature object
+  c3_y input64[64];
+  c3_ws ii;
+  for (ii = 0; ii < 32; ii++){
+    input64[ii] = r[ii] ;
+  }
+  for (ii = 0; ii < 32; ii++){
+    input64[ii + 32] = s[ii] ;
+  }
+
+  c3_ws vi = v[0];
+  vi -= 27;
+  secp256k1_ecdsa_recoverable_signature sig;
+  bzero( (void *)  & sig, sizeof(secp256k1_ecdsa_recoverable_signature)  );
+  c3_ws ret = secp256k1_ecdsa_recoverable_signature_parse_compact(context, // IN:  context
+                                                                & sig,   // OUT: sig
+                                                                input64, // IN:  r/s
+                                                                vi);     // IN:  v
+  if (1 != ret) {
+    fprintf(stderr, "\rsecp jet: crypto package error\n");
+    return u3m_bail(c3__exit);
+  }
+
+  // turn sign into pubkey
+  secp256k1_pubkey pubkey;
+  bzero((void *) &pubkey, sizeof(secp256k1_pubkey));  
+  ret = secp256k1_ecdsa_recover(context,                      // IN: context
+                                & pubkey,                     // OUT: pub key
+                                & sig,                        // IN: signature
+                                (const c3_y *) & has // IN: message has
+                                );
```

Comment:
Here, and a couple lines down in the call to `pubkey_serialize`, it's conventional to put the closing `);` on the same line as the last argument.

Reply:
done